### ISSUE TITLE
Perform download timeout

### ISSUE THREAD

**Issue by Downchuck** [EugeneLoy/edx-archive#8]
Trying to grab an archive of a finished course, getting this unfortunate timeout.

```
Error: WebSocket is not open: readyState 3 (CLOSED)
    at WebSocket.send (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/ws/lib/websocket.js:329:19)
...
    at EdxDownloader.openPage (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/built/Core.js:13:41)
    at /usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/built/EdxDownloader.js:49:37
  -- ASYNC --
    at Browser.<anonymous> (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/puppeteer/lib/helper.js:111:15)
    at EdxDownloader.openPage (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/built/Core.js:13:41)
```

**Comment by EugeneLoy** [comment]
Hi.
1. What course are you downloading? Does it work for other courses (demo course, for example: https://courses.edx.org/courses/edX/DemoX/Demo_Course/course/)?
2. Try raising number of retries by adding `-r <number>` and delay by adding `-d <number>` (defaults are 3 and 1 respectively). If this is not helping, add `--debug` and post full output here.

**Comment by Downchuck** [comment]
Using retry 6, delay 3 -- it's not clear if it actually tried to retry:
```
-c 1 -r 6 -d 3
```

```
Error: WebSocket is not open: readyState 3 (CLOSED)
    at WebSocket.send (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/ws/lib/websocket.js:329:19)
    at WebSocketTransport.send (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/puppeteer/lib/WebSocketTransport.js:60:14)
    at Connection._rawSend (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/puppeteer/lib/Connection.js:86:21)
    at Connection.send (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/puppeteer/lib/Connection.js:72:21)
    at Browser._createPageInContext (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/puppeteer/lib/Browser.js:174:47)
    at BrowserContext.newPage (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/puppeteer/lib/Browser.js:367:26)
    at Browser.newPage (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/puppeteer/lib/Browser.js:166:33)
    at Browser.<anonymous> (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/puppeteer/lib/helper.js:112:23)
    at EdxDownloader.openPage (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/built/Core.js:13:41)
    at /usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/built/EdxDownloader.js:49:37
  -- ASYNC --
    at Browser.<anonymous> (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/puppeteer/lib/helper.js:111:15)
    at EdxDownloader.openPage (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/built/Core.js:13:41)
    at /usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/built/EdxDownloader.js:49:37
    at Observable._subscribe (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/rxjs/internal/observable/defer.js:10:21)
    at Observable._trySubscribe (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/rxjs/internal/Observable.js:44:25)
    at Observable.subscribe (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/rxjs/internal/Observable.js:30:22)
    at Object.subscribeToResult (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/rxjs/internal/util/subscribeToResult.js:12:23)
    at MergeMapSubscriber._innerSub (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/rxjs/internal/operators/mergeMap.js:82:53)
    at MergeMapSubscriber._tryNext (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/rxjs/internal/operators/mergeMap.js:76:14)
    at MergeMapSubscriber._next (/usr/local/Cellar/node/12.12.0/lib/node_modules/edx-archive/node_modules/rxjs/internal/operators/mergeMap.js:59:18)
```

**Comment by EugeneLoy** [comment]
@Downchuck Unfortunately, this is still not enough to figure out what the problem is. In order for me to help you, please, do the following:
1. Tell me what course you are trying to download.
2. Are you able to download demo course (sign up for it before attempting to download)?
3. Add `-c 1 -r 10 -d 10 --debug` switches and post **complete** output of the program (not only stack trace).
4. Repeat (3) multiple times. Does it fail to download specific page every time (if this is not clear from the output, posting complete outputs from multiple runs here will substitute for answer)?

**Comment by Downchuck** [comment]
@EugeneLoy There's nothing useful in the remaining debug information. It's trying to download about 350 pages; it has no capability to resume, so it of course starts from scratch each time.

1. https://courses.edx.org/courses/course-v1:MITx+18.6501x+3T2019/course/
2. Yes
3. There's nothing useful here --  it's simply that it fails after "Downloading task", e.g.
```
Downloading task: https://courses.edx.org//courses/course-v1:MITx+18.6501x+3T2019/courseware/unit1/u1s1_whatisstats/9
{
  id: 'https://courses.edx.org//courses/course-v1:MITx+18.6501x+3T2019/courseware/unit1/u1s1_whatisstats/9',
  name: 'https://courses.edx.org//courses/course-v1:MITx+18.6501x+3T2019/courseware/unit1/u1s1_whatisstats/9',
  url: 'https://courses.edx.org//courses/course-v1:MITx+18.6501x+3T2019/courseware/unit1/u1s1_whatisstats/9',
  index: 35
}
Error: WebSocket is not open: readyState 3 (CLOSED)
```

4. It does not fail on any particular page.

General recommendation is that an on disconnected event needs to be monitored:
https://github.com/puppeteer/puppeteer/issues/4428

Seen also in:
https://github.com/GoogleChrome/rendertron/pull/369

Why it crashes or fails is anyone's guess, but anyone consuming puppeteer seems better off to handle the error.

-Charles

**Comment by EugeneLoy** [comment]
@Downchuck Thanks for the references.
I've implemented handling browser crashes and released it in v1.6.0.
Check it out and let me know if this helped you.

**Comment by Downchuck** [comment]
Yep this addressed the issue -- may have been caused by my laptop having a nap -- but regardless, totally fixed the issue and the restarts worked.

Thank you!
